File: typescript/core/step/Count.ts
import { Lead } from "../Lead.ts";
import { State } from "../State.ts";

export async function count(state: State): Promise<[State, Lead | null]> {
  console.log(`Current count: ${state.count}`);
  await new Promise((resolve) => setTimeout(resolve, 1000));
  state.count++;

  if (Math.random() < 0.5) {
    return [state, Lead.COUNT];
  } else {
    return [state, Lead.STOP];
  }
}

File: kotlin/src/core/step/Count.kt
package core.step

import Step
import core.Lead
import kotlinx.coroutines.delay
import kotlin.random.Random

object Count : Step<Int, Lead> {
    override suspend fun resolve(state: Int): Pair<Int, Lead?> {
        println("Count: $state")
        delay(1000)
        val newState = state + 1
        return newState to when {
            Random.Default.nextDouble() > 0.5 -> Lead.COUNT
            else -> Lead.STOP
        }
    }
}
